crates/service-protocol/src/discovery.rs

```diff
@@ -152,26 +165,81 @@ impl ServiceDiscovery {
+                ));
+            }
+            None => {
+                return Err(DiscoveryError::BadResponse(format!("No content type header was specified. Expected '{APPLICATION_JSON:?}' content type.").into()))
             }
         }
 
         // Parse the response
-        let response: schema::Deployment =
+        let response: schema::Endpoint =
             serde_json::from_slice(&body).map_err(|e| DiscoveryError::Decode(e, body))?;
 
-        let protocol_type = match response.protocol_mode {
+        Self::create_discovered_metadata_from_endpoint_response(response)
+    }
+
+    fn create_discovered_metadata_from_endpoint_response(
+        endpoint_response: schema::Endpoint,
+    ) -> Result<DiscoveredMetadata, DiscoveryError> {
+        let protocol_type = match endpoint_response.protocol_mode {
             Some(schema::ProtocolMode::BidiStream) => ProtocolType::BidiStream,
             Some(schema::ProtocolMode::RequestResponse) => ProtocolType::RequestResponse,
             None => {
-                return Err(DiscoveryError::BadResponse("missing protocol mode"));
+                return Err(DiscoveryError::BadResponse("missing protocol mode".into()));
             }
         };
 
+        // Sanity checks for the service protocol version
+        if endpoint_response.min_protocol_version <= 0
+            || endpoint_response.min_protocol_version > MAX_SERVICE_PROTOCOL_VERSION_VALUE as i64
+        {
+            return Err(DiscoveryError::BadResponse(
+                format!(
+                    "min protocol version must be in [1, {}]",
+                    MAX_SERVICE_PROTOCOL_VERSION_VALUE
+                )
+                .into(),
+            ));
+        }
+
+        if endpoint_response.max_protocol_version <= 0
+            || endpoint_response.max_protocol_version > MAX_SERVICE_PROTOCOL_VERSION_VALUE as i64
+        {
+            return Err(DiscoveryError::BadResponse(
+                format!(
+                    "max protocol version must be in [1, {}]",
+                    MAX_SERVICE_PROTOCOL_VERSION_VALUE
+                )
+                .into(),
+            ));
+        }
+
+        if endpoint_response.min_protocol_version > endpoint_response.max_protocol_version {
+            return Err(DiscoveryError::BadResponse(
+                format!("Expected min protocol version to be <= max protocol version. Received min protocol version '{}', max protocol version '{}'", endpoint_response.min_protocol_version, endpoint_response.max_protocol_version).into(),
+            ));
+        }
+
+        let min_version = endpoint_response.min_protocol_version as i32;
+        let max_version = endpoint_response.max_protocol_version as i32;
+
+        if !protocol::Version::is_supported(min_version, max_version) {
+            return Err(DiscoveryError::UnsupportedServiceProtocol {
+                min_version,
+                max_version,
+            });
+        }
```

Comment:
What still can happen is if the server removes a supported version that some registered endpoints can no longer be served. So ideally, the server never returns supported protocol versions.

Reply:
would the invoker still be able to do this check?